File: typescript/creational/abstract-factory/WoodenDoorFactory.ts
import {DoorFactory} from './DoorFactory';
import {Door} from './Door';
import {WoodenDoor} from './WoodenDoor';
import {DoorFittingExpert} from './DoorFittingExpert';
import {Carpenter} from './Carpenter';

// Wooden factory to return carpenter and wooden door
export class WoodenDoorFactory implements DoorFactory {
  public makeDoor():Door {
    return new WoodenDoor();
  }

  public makeFittingExpert():DoorFittingExpert {
    return new Carpenter();
  }
}

File: ruby/creational/abstract-factory/WoodenDoorFactory.rb
require_relative 'DoorFactory'
require_relative 'WoodenDoor'
require_relative 'Carpenter'

class WoodenDoorFactory include DoorFactory
  def make_door
    return WoodenDoor.new
  end

  def make_fitting_expert
    return Carpenter.new
  end
end
